test/cql/query_bounds_test.result

```diff
@@ -0,0 +1,442 @@
+			"a" : "0",
+			"b" : "0",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b, c) <= SCYLLA_CLUSTERING_BOUND (2, 0);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "2",
+			"c" : "0"
+		}
+	]
+}
+
+
+SELECT * FROM foo WHERE a=0 AND (b) > (1);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "2",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) > SCYLLA_CLUSTERING_BOUND (1);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "0",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) > (2);
+{
+	"rows" : null
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) > SCYLLA_CLUSTERING_BOUND (2);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "0"
+		},
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "1"
+		},
+		{
+			"a" : "0",
+			"b" : "0",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) >= (2);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "2",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) >= SCYLLA_CLUSTERING_BOUND (2);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "2",
+			"c" : "0"
+		},
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "0"
+		},
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "1"
+		},
+		{
+			"a" : "0",
+			"b" : "0",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) < (1);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "0",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) < SCYLLA_CLUSTERING_BOUND (1);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "2",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) < (2);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "0"
+		},
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "1"
+		},
+		{
+			"a" : "0",
+			"b" : "0",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) < SCYLLA_CLUSTERING_BOUND (2);
+{
+	"rows" : null
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) <= (2);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "2",
+			"c" : "0"
+		},
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "0"
+		},
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "1"
+		},
+		{
+			"a" : "0",
+			"b" : "0",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) <= SCYLLA_CLUSTERING_BOUND (2);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "2",
+			"c" : "0"
+		}
+	]
+}
+
+
+
+
+SELECT * FROM foo WHERE a=0 AND (b) > (1) AND (b, c) <= (2, 0);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "2",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) > SCYLLA_CLUSTERING_BOUND (1) AND (b, c) <= SCYLLA_CLUSTERING_BOUND (0, 0);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "0",
+			"c" : "0"
+		}
+	]
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) > (2) AND (b, c) <= (2, 1);
+{
+	"rows" : null
+}
+
+SELECT * FROM foo WHERE a=0 AND (b) > SCYLLA_CLUSTERING_BOUND (2) AND (b, c) <= SCYLLA_CLUSTERING_BOUND(1, 1);
+{
+	"rows" : 
+	[
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "0"
+		},
+		{
+			"a" : "0",
+			"b" : "1",
+			"c" : "1"
+		}
+	]
+}
+
```

Comment:
Please add some negative tests, where non-clustering keys are used or when they are used in the wrong order.

Now I see that SCYLLA_CLUSTERING_BOUND() is used in the RHS, not LHS. Does it not make sense to use it on the LHS? Maybe it should be around both.

Reply:
Why? The way I see it, it tags "values" (or value references - vars) as being "special". And from a pure ANTLR perspective, it is much easier to get it in on RHS. Otherwise I have to more or less copy (or refactor/rewrite) the whole restriction match grammar.

Reply:
What would be accomplished by forcing the user to type it twice?  Once seems enough.

Reply:
I see it as making the relational operator special. Since the relational operator's semantics are type specific, the function-style casting are enough to change it.

When the user types it twice, the semantics are: we compare two clustering keys. When the user types it once, for the RHS, the semantics are: we change the LHS type based on the RHS type.

But twice (or only RHS) allows this too:

```sql
WHERE SCYLLA_CLUSTERING_BOUND(ck1, ck2, ck3) > ?
```

And so allow a bound tuple on the RHS, not just bound elements.

Reply:
The syntax allows a bound tuple. It is not actually a function as such, it is a qualifying keyword. I should maybe change the docs to state this, but "function" seemed like it would more discourage over usage of it. And again: internal stuff.

Reply:
Or rather, I think it allows a bound tuple, assuming the parser can deal with it. Not 100% on this. Because it is not a use case I care about.

Reply:
Isn't a bound tuple the main use case? or is the plan to use SCYLLA_CLUSTERING_BOUND(?, ?, ?) and bind NULLs to trailing elements when not specified? Does the code support this?

Reply:
The uage in sstableloader is to bind as many columns + vars/values as exist in the bounds read from sstable, which effectively means nulls for the traiing ones. Though not expressed explicitly. While this does increase the number of statements prepared compared to for example always have full clustering column set included, it is probably not that big a deal. The (admittedly small) sample set of data to provoke this suggests that we get regular patterns of bounds for range expressions (i.e. tombstones), and inserts/updates are always equal comparisons.

Reply:
Ok. I regard this as semi-internal (it can be used by other migration applications, and we will have to support it) so it merits extra scrutiny, but we can come up with a new one if the need arises.